plutus-use-cases/src/Language/PlutusTx/Coordination/Contracts/Currency.hs

```diff
@@ -0,0 +1,171 @@
+import           Ledger.Validation         (TxHash)
+import qualified Ledger.Validation         as V
+import qualified Ledger.Value.TH           as Value
+import           Ledger                    as Ledger hiding (to)
+import           Ledger.Value              (TokenName, Value, mkTokenName)
+import           Wallet.API                as WAPI
+
+import qualified Language.PlutusTx.Coordination.Contracts.PubKey as PK
+
+data Currency = Currency
+                    { curRefTransactionOutput :: (TxHash, Int)
+                    -- ^ Transaction input that must be spent when
+                    --   the currency is forged.
+                    , curRefAmount            :: Int
+                    -- ^ How much of the currency can be forged
+                    , curRefTokenName         :: TokenName
+                    -- ^ Token name
+                    }
+
+P.makeLift ''Currency
+
+mkCurrency :: TxOutRef -> Int -> String -> Currency
+mkCurrency (TxOutRefOf h i) amt n = 
+    Currency
+        { curRefTransactionOutput = (V.plcTxHash h, i)
+        , curRefAmount            = amt
+        , curRefTokenName         = mkTokenName n
+        }
+
+curValidator :: Currency -> ValidatorScript
+curValidator cur =
+    ValidatorScript (Ledger.applyScript mkValidator (Ledger.lifted cur)) where
+        mkValidator = Ledger.fromCompiledCode ($$(P.compile [||
+            let validate :: Currency -> () -> () -> V.PendingTx -> ()
+                validate (Currency (refHash, refIdx) amt token) () () p = 
+                    let 
+                        ownSymbol = $$(V.ownCurrencySymbol) p
+                        forged = $$(V.valueForged) p
+                        expected = $$(Value.singleton) ownSymbol token amt 
+
+                        -- True if the pending transaction forges the amount of
+                        -- currency that we expect
+                        forgeOK =
+                            let v = $$(Value.eq) expected forged
+                            in $$(P.traceIfFalseH) "Value forged different from expected" v
+
+                        -- True if the pending transaction spends the output
+                        -- identified by @(refHash, refIdx)@
+                        txOutputSpent = 
+                            let v = $$(V.spendsOutput) p refHash refIdx
+                            in  $$(P.traceIfFalseH) "Pending transaction does not spend the designated transaction output" v
+
+                    in
+                        if $$(P.and) forgeOK txOutputSpent
+                        then ()
+                        else $$(P.error) ($$(P.traceH) "Invalid forge" ())
+            in
+                validate
+            ||]))
+
+-- | Currency data script (unit value).
+curDataScript :: DataScript
+curDataScript = DataScript $ Ledger.lifted ()
+
+-- | Currency redeemer (unit value).
+curRedeemer :: RedeemerScript
+curRedeemer = RedeemerScript $ Ledger.lifted ()

```

Comment:
I'm not sure how much it adds to give top-level bindings to all these unit scripts we have all over the place - can't we just inline them?